### ISSUE TITLE
Trigger Build On GitHub Repo Webhook

### ISSUE THREAD

**Issue by TomGranot** [concourse/concourse#4955]
Basically migrating https://github.com/telia-oss/github-pr-resource/issues/183 here, since I'm not sure it will get much reading there:

Hiya!

Deeply considering the move to Concourse for a while now, and I'm currently constructing a copy of our existing CI setup for one of our projects to give it a spin.

Our flow today is as follows in (using GitHub + Azure Pipelines):

1. PR is opened
2. Since PR includes at least one commit, PR triggers CI pipeline
3. If any step fails, CI pushes a status to GitHub checks
4. If pipeline completed, CI pushes a status to GitHub checks

I've managed to replicate the last 3 steps in Concourse in full, quite fast and using the command line on a local Concourse setup which is super-duper-awesome (much, much faster the re-doing CI configurations in Azure). 

Step 1, however, remains semi-unsolved. I've got two options to know that a PR has been opened:

1. **poll** - Use `every_which` in my resource declaration to poll the relevant resource every so-and-so seconds (where the resource's `resource-type` is `github-pr-resource`, based on https://github.com/telia-oss/github-pr-resource).
2. **push** - Wait for a webhook to come from GitHub whenever a new `push` or `pull_request` event have been created on the repo.

The `poll` version works fine, but has the obvious problem of waiting for the timeout to expire and of over-polling the GitHub API which may cause problems at the future.

The `push` version **does not work**, or at least not how I would expect it to work. I've looked at https://github.com/concourse/concourse/issues/3411 and it seems this is a similar issue to what I'm facing: I'd like the webhook to trigger the **pipeline**, no the **check** of the resource.

Some help would be much appreciated - I've set the webhook on the repo with the correct privileges, set the personal access token with the correct privileges as well, and can see that `POST`ing to the relevant webhook returns a 201 with the response I'd expect from concourse. 

My config (based on the example from the documentation):

```
resource_types:
- name: pull-request-base-resource
  type: docker-image
  source:
    repository: teliaoss/github-pr-resource

resources:
- name: pull-request
  type: pull-request-base-resource
  check_every: 10s
  webhook_token: <WEBHOOK_TOKEN> 
  source:
    repository: <MY_REPO>
    access_token: <MY_PERSONAL_ACCESS_TOKEN>
    base_branch: <MY_BRANCH>
jobs:
- name: test
  plan:
  - get: pull-request
    trigger: true
    version: latest # Tried with every as well...
  - put: pull-request
    params:
      path: pull-request
      status: pending
  - task: unit-test
    config:
      platform: linux
      image_resource:
        type: docker-image
        source: {repository: alpine/git, tag: "latest"}
      inputs:
        - name: pull-request
      run:
        path: /bin/sh
        args:
          - -xce
          - |
            cd pull-request
            git log --graph --all --color --pretty=format:"%x1b[31m%h%x09%x1b[32m%d%x1b[0m%x20%s" > log.txt
            cat log.txt
    on_failure:
      put: pull-request
      params:
        path: pull-request
        status: failure
  - put: pull-request
    params:
      path: pull-request
      status: success
```

**Comment by pivotal-jamie-klassen** [comment]
@tomgs I [commented](https://github.com/telia-oss/github-pr-resource/issues/183#issuecomment-570207732) on the upstream issue, asking for more details. I asked there because, after carefully reading your description and the similar issue you linked (thanks for doing such thoughtful research!), I still want to conclude that it's either

1. a problem with the pull-request resource type iteself, or
1. a problem with your configuration

This is because the difference between "triggering the pipeline" and "checking the resource" as you've described them is actually pretty immaterial (at least when your job has `trigger: true` on a `get` step involving the resource in question).

What I mean by this is that you could probably cook up a way to have github's webhook make a request to the [`CreateJobBuild`](https://github.com/concourse/concourse/blob/master/atc/routes.go#L135) endpoint (which is what running `fly trigger-job` or clicking the + button in the UI does) instead of `CheckResourceWebHook` and I strongly suspect the behaviour would be the same -- a resource check would be queued up, and (once it finished) a build would run using the latest version that was discovered as an input.

**Comment by tomgs** [comment]
@pivotal-jamie-klassen added a through response on the other thread - thanks for looking out!

Following this:

> What I mean by this is that you could probably cook up a way to have github's webhook make a request to the CreateJobBuild endpoint (which is what running fly trigger-job or clicking the + button in the UI does) instead of CheckResourceWebHook and I strongly suspect the behaviour would be the same -- a resource check would be queued up, and (once it finished) a build would run using the latest version that was discovered as an input.

Am I to understand the logic I'm trying to implement is actually not possible using Concourse? I'm wondering out loud because this feels like a very common workflow for dev teams all over the place, and since Concourse is quite popular I assumed there'd be a more "straightforward" way of doing this.

Or am I misunderstanding?

And thanks again for your thorough responses! :)

**Comment by vito** [comment]
Hey, thanks for the feedback! I think this is an important issue but there's also a lot of nuance to it. Apologies in advance for my wordy response. :sweat_smile: The tl;dr is that what you're precisely asking for (triggering a build through a webhook) is not possible, but I think there are other ways we can achieve what you want.

Firstly: if the PR flow setup feels awkward/obtuse, you're completely right. This is the primary focus of our [v10 roadmap](https://blog.concourse-ci.org/core-roadmap-towards-v10/), which is itself a pretty wordy blog post, but it's worth a read if you have time!

> The `push` version **does not work**, or at least not how I would expect it to work. I've looked at #3411 and it seems this is a similar issue to what I'm facing: I'd like the webhook to trigger the **pipeline**, no the **check** of the resource.

One thing probably worth clarifying here: in Concourse there's no such thing as triggering a pipeline, and there is no such thing as e.g. 'pipeline runs'. Pipelines are just sets of jobs, and they are all independently scheduled. New builds are created by the scheduler through changes to a job's dependent resources - `get` steps - with `trigger: true` configured. This is why hooks are tied to resources, not jobs. (In the future I would like to have the hooks just point at the entire Concourse and have Concourse figure out which resources need to be `check`ed, but that's a topic for another day.)

Right now the delay is because the end-to-end flow (PR created -> status checks appear) depends on a lot of asynchronous, eventually-consistent moving parts:

1. The hook queues up a `check`, and once new versions are found they are saved into the database.
1. Up to 10 seconds later, the scheduler runs, notices the new versions, and creates at most one pending build per job.
1. The build starter (which runs immediately after the scheduler) notices a newly queued build and starts it if possible, using the latest available inputs. If the job is configured with `serial: true` or `max_in_flight: N`, the build may not be able to start for an arbitrary amount of time (prior builds have to complete).
1. The above two steps will happen in lock-step, because the scheduler only determines a single set of inputs at a time, and `version: every` only moves on to the next version once a build has started with the previous one. This means that if there are 3 new PRs it'll take at least 3 schedule + start ticks (a minimum of 30 seconds) to chew through them all, resulting in a lengthy delay for the last PR processed.
1. The build starts, and the `put` step which sets the status check to 'pending' runs. This itself takes a lot of steps:
    1. If your `put` step isn't the first thing in the plan, all the steps before it will have to finish. This is arbitrary added time.
    1. Picking a worker for the `put` step container. This shouldn't take very long.
    1. Creating the input volumes for the `put` container, streaming the contents over if they're on a different worker. This can take a long time if your inputs are large. There may also be more inputs than necessary if the `put` step is preceded by a bunch of `get` steps besides the PR resource itself, so you may want to consider explicitly configuring [`inputs`](https://concourse-ci.org/put-step.html#put-step-inputs) to only list the PR resource.
    1. Finally, creating the `put` step container and running it. At this point your status checks finally arrive.

This is slower than other CI systems which are more deeply integrated with GitHub. They probably interpret the webhook payload and immediately create a build associated directly to the commit, and report that build's status check back to GitHub. It's nice and snappy but it's also tightly coupled to GitHub - something Concourse tries real hard to avoid.

The advantage of the current approach is technical simplicity/consistency. It keeps everything in Concourse working in the same way - the resource's `check` remains the source of truth for the pipeline, and resource types aren't required to handle hook payloads from various services - they just have to implement `check`, and the webhooks are purely in service of speeding up the usual eventually-consistent resource `check`s. This reduces the scope of work for resource type authors and prevents inconsistencies in the handling of webhooks and the periodic `check`. But it's obviously not fast enough.

Taking a step back, I don't think these technical details are actually what you care about. From reading the other issue it sounds like you really just want things to happen more quickly, i.e. having the status checks appear on your PR as soon as possible in order to minimize the amount of time where the status checks can be bypassed. And also to just have snappier UX - a valid reason on its own.

If we frame the problem this way I think there are way more options for us to improve:

1. We can eliminate that initial 10 second gap by scheduling downstream jobs immediately after the versions are saved. This will be pretty easy with v6.0's new scheduler.
1. We can have the scheduler create all the pending builds at once, with their inputs determined already, to eliminate the lock-step schedule + start behavior.
1. Once we have support for [notifications](https://blog.concourse-ci.org/core-roadmap-towards-v10/#rfc-28-notification-resources) we can model the GitHub status checks as notifications which are run by Concourse when a build is pending, started, and finished. This will eliminate the `put` step entirely.

In this world, the end-to-end flow would look something like this:

1. The hook queues up a `check`, and once new versions are found they are saved into the database.
1. After saving new versions, Concourse internally requests scheduling of all jobs downstream of the resource and "kicks" the scheduler to take action *immediately*, short-circuiting the polling interval.
1. The scheduler creates all of the pending builds in one tick instead of in lock-step with starting them, each build associated to the appropriate PR.
1. Some sort of notification mechanism in Concourse automatically emits a "pending" notification for each pending build. At this point your status checks will arrive.

This flow would eliminate the eventually-consistent intervals and should really shorten the time it takes to see a status check appear in GitHub. These improvements will also have greater impact outside the scope of the PR flow - in general *all* scheduling will be much quicker, and first-class support for notifications will really clean up pipeline configs.

I hope this helps! I'm sorry we don't really have a quick solution here, other than watching out for the pitfalls I mentioned in the current 'end-to-end' flow and implementing the suggested mitigations (e.g. `inputs:` on the `put` step). This is definitely a high priority issue, but it'll take a while to get there.

The notification mechanism is probably the biggest missing piece. I hope to draft an RFC for it soon so there's something more concrete to point to. There's a placeholder here if you'd like to monitor the progress: https://github.com/concourse/rfcs/pull/28 - and who knows, maybe someone from the community will tackle it once the RFC is done if we don't get to it in time ourselves. :crossed_fingers:

**Comment by tomgs** [comment]
Hey @vito - wordy responses are better than brief responses are better than no response... I wholeheartedly appreciate you taking the time to write this, and hope you have an awesome weekend! 

As per your points:

> The tl;dr is that what you're precisely asking for (triggering a build through a webhook) is not possible, but I think there are other ways we can achieve what you want.

I've been messing around with Concourse for the past few days and I see now that my thinking was a bit upside down. I'm trying to replicate what we had with Azure Pipelines in Concourse, and that's lopsided. A better approach would be to re-model our CI flow to fit "idiomatic" Concourse - I'm currently trying to figure out whether we get a bang for our (dev-time) buck here. Not convinced yet, to be honest, but really rooting for a solution to be found.

> One thing probably worth clarifying here: in Concourse there's no such thing as triggering a pipeline, and there is no such thing as e.g. 'pipeline runs'. Pipelines are just sets of jobs, and they are all independently scheduled.

Exactly my point from before. In Azure Pipelines, if a commit is made in a PR, the **pipeline** is triggered, setting off all the jobs in it. I was trying to do that in Concourse, which is not exactly what I should do - instead, I'm better off setting up the correct resource for each job, and make sure that resource is triggered correctly with the webhook.

> Right now the delay is because the end-to-end flow (PR created -> status checks appear) depends on a lot of asynchronous, eventually-consistent moving parts:

Ah! OK, I get the flow now. To be perfectly honest, this is a bit more involved than I hoped it would be for something that is a rather "traditional" workflow, but again - I understand that's where my goals deviate from Concourse's (at least in theory).

To me, this: 

> Once we have support for notifications we can model the GitHub status checks as notifications which are run by Concourse when a build is pending, started, and finished. This will eliminate the put step entirely.

Would be the way I would like to have things happen. While keeping the original "flow" of Concourse consistent for all users, you'll be allowing integrations for specific platforms much more easily. The Checks screen at GitHub (or any other hosted VCS server) assume that such a logic is derived at build time in the CI server - and thus many CI servers accommodate that view (at least Travis, Circle and Azure do, from my experience).

> I hope this helps! I'm sorry we don't really have a quick solution here, other than watching out for the pitfalls I mentioned in the current 'end-to-end' flow and implementing the suggested mitigations (e.g. inputs: on the put step). This is definitely a high priority issue, but it'll take a while to get there.

Don't be. You can't fit all the birds in one tree without breaking some eggs along the way (or whatever the saying is, I'm not a native English speaker). I'll give the `put` step `inputs` a spin and see if that clears things up on my end, and report back. 

In the meantime, I'm going to continue giving Concourse a run for its money while taking this issue under advisement. Monitoring https://github.com/concourse/rfcs/pull/28 in the meantime. BTW - I'm not sure what the methodology is here - we can either close this awaiting for the RFC to be implemented or keep it open, since the problem is not actually solved.

And again, thanks for all your help @vito and @pivotal-jamie-klassen! It's really awesome to see such great support being offered for the product. Hope this works out!

**Comment by afonsoma9** [comment]
I am trying to configure the concourse to execute a pipeline at each pull request and although the webhook is being sent and answered successfully the pipeline does not start.

### Request
![Captura de tela de 2020-01-17 16-36-16](https://user-images.githubusercontent.com/38052959/72641433-f93a5680-3948-11ea-82b1-78019c3b29b2.png)

### Response
![Captura de tela de 2020-01-17 16-36-32](https://user-images.githubusercontent.com/38052959/72641439-fccddd80-3948-11ea-8803-7ccba2f497f7.png)

### My Compose Configuration
```yml
version: "3"

services:
  gitea-app:
    image: gitea/gitea:latest
    container_name: gitea-app
    environment:
      - USER_UID=1000
      - USER_GID=1000
      - ROOT_URL=http://localhost:3000
      - SSH_DOMAIN=mydomain.com
    restart: always
    volumes:
      - ./volumes/gitea_app:/data
    ports:
      - "3000:3000"
      - "22:22"
    networks:
      - appnet

  gitea-db:
    image: postgres:alpine
    container_name: gitea-db
    ports:
      - 5440:5432
    restart: always
    volumes:
      - ./volumes/gitea_db:/var/lib/postgresql/data
    environment:
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
      - POSTGRES_DB=gitea
    networks:
      - appnet

  concourse-db:
    image: postgres
    environment:
      POSTGRES_DB: concourse
      POSTGRES_PASSWORD: concourse_pass
      POSTGRES_USER: concourse_user
      PGDATA: /database
    networks:
      - appnet

  concourse:
    image: concourse/concourse
    command: quickstart
    privileged: true
    depends_on: [concourse-db]
    ports: ["8080:8080"]
    environment:
      CONCOURSE_POSTGRES_HOST: concourse-db
      CONCOURSE_POSTGRES_USER: concourse_user
      CONCOURSE_POSTGRES_PASSWORD: concourse_pass
      CONCOURSE_POSTGRES_DATABASE: concourse
      CONCOURSE_EXTERNAL_URL: http://localhost:8080
      CONCOURSE_ADD_LOCAL_USER: admin:admin
      CONCOURSE_MAIN_TEAM_LOCAL_USER: admin
      CONCOURSE_WORKER_BAGGAGECLAIM_DRIVER: naive
    networks:
      - appnet

volumes:
  gitea-app:
  gitea-db:

networks:
  appnet:
    external: true
```
### My Pipeline Configuration
```yml
resources:
- name: principal
  type: git
  check_every: 10s
  webhook_token: principal
  source:
    uri: "http://gitea-app:3000/afonsoma9/principal.git"
    branch: develop


jobs:
- name: test
  plan:
    - get: principal
      trigger: true
    - task: run-tests
      config:
        platform: linux
        image_resource:
          type: docker-image
          source:
            repository: circleci/node
            tag: '12.14.1-browsers'
        run:
          path: sh
          args:
                - -c
                - |
                  export NG_CLI_ANALYTICS=ci
                  echo "Node Version: $(node --version)"
                  echo "NPM Version: $(npm --version)"
                  sudo chown -R $USER principal
                  cd principal
                  npm install
                  ls -la
                  whoami
                  npm run test-headless --unsafe-perm
        inputs:
          - name: principal
```
Can anyone help?

**Comment by tooolbox** [comment]
@afonsoma9 maybe you want to check out #4681 ?

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
